sweet/benchmarks/cockroachdb/main.go

```diff
@@ -0,0 +1,611 @@
+	"github.com/pkg/errors"
+	"golang.org/x/benchmarks/sweet/benchmarks/internal/driver"
+	"golang.org/x/benchmarks/sweet/benchmarks/internal/server"
+	"golang.org/x/benchmarks/sweet/common/diagnostics"
+	"os"
+	"os/exec"
+	"regexp"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+)
+
+const (
+	// Arbitrarily chosen to match the cockroachdb default.
+	basePort = 26257
+	// The percentage of memory to allocate to the cache.
+	cacheSize = "0.25"
+)
+
+type config struct {
+	host           string
+	cockroachdbBin string
+	tmpDir         string
+	benchName      string
+	isProfiling    bool
+	short          bool
+	procsPerInst   int
+	bench          *benchmark
+}
+
+var cliCfg config
+
+func init() {
+	driver.SetFlags(flag.CommandLine)
+	flag.StringVar(&cliCfg.host, "host", "localhost", "hostname of cockroachdb server")
+	flag.StringVar(&cliCfg.cockroachdbBin, "cockroachdb-bin", "", "path to cockroachdb binary")
+	flag.StringVar(&cliCfg.tmpDir, "tmp", "", "path to temporary directory")
+	flag.StringVar(&cliCfg.benchName, "bench", "", "name of the benchmark to run")
+	flag.BoolVar(&cliCfg.short, "short", false, "whether to run a short version of this benchmark")
+}
+
+type cockroachdbInstance struct {
+	name     string
+	sqlPort  int // Used for intra-cluster communication.
+	httpPort int // Used to scrape for metrics.
+	cmd      *exec.Cmd
+	output   bytes.Buffer
+}
+
+func clusterAddresses(instances []*cockroachdbInstance) string {
+	var s []string
+	for _, inst := range instances {
+		s = append(s, inst.sqlAddr())
+	}
+	return strings.Join(s, ",")
+}
+
+func launchSingleNodeCluster(cfg *config) ([]*cockroachdbInstance, error) {
+	var instances []*cockroachdbInstance
+	instances = append(instances, &cockroachdbInstance{
+		name:     "roach-node",
+		sqlPort:  basePort,
+		httpPort: basePort + 1,
+	})
+	inst := instances[0]
+
+	// `cockroach start-single-node` handles both creation of the node
+	// and initialization.
+	inst.cmd = exec.Command(cfg.cockroachdbBin,
+		"start-single-node",
+		"--insecure",
+		"--listen-addr", inst.sqlAddr(),
+		"--http-addr", inst.httpAddr(),
+		"--cache", cacheSize,
+		"--store", fmt.Sprintf("%s/%s", cfg.tmpDir, inst.name),
+		"--logtostderr",
+	)
+	inst.cmd.Env = append(os.Environ(),
+		fmt.Sprintf("GOMAXPROCS=%d", cfg.procsPerInst),
+	)
+	inst.cmd.Stdout = &inst.output
+	inst.cmd.Stderr = &inst.output
+	if err := inst.cmd.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start instance %q: %v", inst.name, err)
+	}
+	return instances, nil
+}
+
+func launchCockroachCluster(cfg *config) ([]*cockroachdbInstance, error) {
+	if cfg.bench.nodeCount == 1 {
+		// Use `cockroach start-single-node` instead for single node clusters.
+		return launchSingleNodeCluster(cfg)
+	}
+	var instances []*cockroachdbInstance
+	for i := 0; i < cfg.bench.nodeCount; i++ {
+		instances = append(instances, &cockroachdbInstance{
+			name:     fmt.Sprintf("roach-node-%d", i+1),
+			sqlPort:  basePort + 2*i,
```

Comment:
Same comment as above... let's double-check if [basePort, basePort+nodeCount*2) is available.